### DISCUSSION TITLE
Handling routing and reactive state (also plugins)

### DISCUSSION THREAD

**Discussion by Klavionik** [Textualize/textual#3187] [Q&A]
Hello! First of all, thank you for Textual! I've been following this project since JetBrains webinar in 2021, and find it really promising. 

Lately I decided to try and build a multiplayer Battleship the Game with terminal UI (using Textual, of course). I have experience building web apps with SPA frameworks like Vue.js and React, and AFAIR Textual is inspired by these.

There is a couple of questions I had right in the beginning, and I guess it will be better to ask you about it.

1. Application state is a major topic in UI development. Textual has a notion of reactive state, but I guess it works a bit differently than React, right? I mean, I cannot pass reactive vars down like props and expect widgets to update. I have to manually watch for variable update, query for required widget and set its attribute if I need. Do you have plans to implement something like real props? I tried to port some reactivity principles from Vue.js to Python and try it with Textual. It works, but it seems a bit hacky and verbose. You really would need to make changes to the metaclass that creates widgets to make it more comfortable to use. But still, it seems doable.
2. I've been trying to figure out how do I switch "pages" during the game (like, the initial page is LoginPage with welcome text and login/password input, after login you get to the game lobby and so on). At first I thought that I should use the Screens API, but it doesn't seem to be the right choice. Screens got pushed on stack, preserving state, so I could get back to them later. And I don't really need to preserve the state of LoginPage after login. So the only way I see right now is to manually query for some kind of "page" widget and unmount/mount to "go" to the other page. Maybe you could give me an advice on that?
3. Also, do you plan to implement any plugin API? I guess the power of SPA frameworks lies in their communities. Core devs can focus on the framework and offload a lot of work to community to build. I have this question because I tried to implement some generic router for Textual (see q. 2), and I'm still figuring out how to propely inject it into the app (right now I'm attaching it as a class variable and using `__get__` to bind it to the app).

Thank you for attention, and sorry if my questions are a bit unclear. I tried to keep this text as short as possible. I would be happy to give more details.

**Comment by davep** [comment]
Regarding question 2: given your example of a login page, it feels like the `Screen` API and pushing and popping would be ideal there. You'd push the login screen, let it do its work, and then pop it once it has done; no state being preserved.

**Reply by Klavionik** [reply]
Thanks a lot for the advice! Sad to say, I didn't think about it. I guess it would work just fine, although using modal screens can make it a bit more complex (I would have to know to pop twice before pushing the next screen).

What would you say about reactive props or a plugin system? I know you have some sort of reactive state features on the roadmap. Is it the same I'm talking about?

**Reply by davep** [reply]
I'm not aware of any details for those questions so can't confidently answer them.
